{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "0a8d2732-9863-4787-a5c2-9de390bfe675",
   "metadata": {},
   "source": [
    "# Intersecting Haplotype Motif Annotations with CHM13\n",
    "We use the CHM13 reference genome Non-B Annotations and intersect ALL Haplotypes with the reference to understand which Motifs are shared by Haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a3eba5d2-dcd8-43dc-96bc-830ca5c74d4f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Top 10 Most Shared Positive G-Quadruplex Regions:\n",
      "       chrom      start        end  unique_haplotype_count\n",
      "22539  chr12  133005163  133011865                      49\n",
      "89108   chr8  144820876  144824591                      47\n",
      "13132  chr11     692395     693434                      43\n",
      "58316  chr20     321487     322359                      42\n",
      "22297  chr12  131961785  131964685                      38\n",
      "71594   chr4    1018653    1019280                      36\n",
      "8916   chr10     511390     512817                      34\n",
      "35183  chr17     941125     942260                      34\n",
      "21705  chr12  121740473  121744476                      31\n",
      "13510  chr11    1465130    1466290                      31\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def process_intersections(bed_file):\n",
    "    # Read BED file into a DataFrame\n",
    "    df = pd.read_csv(bed_file, sep='\\t', header=None, names=['chrom', 'start', 'end', 'stability1', 'filename', 'chrom2', 'start2', 'end2', 'stability2'])\n",
    "    \n",
    "    # Drop duplicate filenames for the same region, considering only the relevant columns\n",
    "    df_unique = df.drop_duplicates(subset=['chrom', 'start', 'end', 'filename'])\n",
    "    \n",
    "    # Group by chrom, start, and end columns and count unique filenames\n",
    "    grouped = df_unique.groupby(['chrom', 'start', 'end'])['filename'].nunique().reset_index(name='unique_haplotype_count')\n",
    "    \n",
    "    # Sort the grouped DataFrame by unique_haplotype_count in descending order\n",
    "    sorted_grouped = grouped.sort_values(by='unique_haplotype_count', ascending=False)\n",
    "    \n",
    "    return sorted_grouped\n",
    "\n",
    "# Main script\n",
    "bed_file_path = '/home/alextu/scratch/results/bed_files/chm13/quadron_haplotype_intersections_chm13_query/verkko_batch123_chm13_quadron_posstrand_intersected_output.bed'\n",
    "grouped_intersections = process_intersections(bed_file_path)\n",
    "\n",
    "# Print the top 10 most shared G-quadruplex regions\n",
    "print(\"Top 10 Most Shared Positive G-Quadruplex Regions:\")\n",
    "print(grouped_intersections.head(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5e15a802-2376-4509-9ce5-249032c1f26e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       chrom      start        end  unique_haplotype_count  \\\n",
      "93      chr1     469936     470018                       9   \n",
      "114     chr1     486907     486944                       6   \n",
      "140     chr1     520317     520447                       5   \n",
      "160     chr1     565868     565926                       5   \n",
      "165     chr1     567815     567861                       5   \n",
      "1335    chr1    5954780    5954834                       5   \n",
      "1352    chr1    5996534    5996559                       5   \n",
      "8879   chr10      34292      34337                       5   \n",
      "13227  chr11     889737     889788                       5   \n",
      "13842  chr11    2185143    2185175                       5   \n",
      "18724  chr11  132753385  132753412                       5   \n",
      "18746  chr11  133367485  133367514                       5   \n",
      "26188  chr14   95998057   95998119                       5   \n",
      "30169  chr16     988458     988576                       5   \n",
      "35165  chr17     865467     865689                       5   \n",
      "35210  chr17    1069220    1069282                       5   \n",
      "35812  chr17    4750240    4750269                       5   \n",
      "38078  chr17   41083456   41083580                       5   \n",
      "38465  chr17   44699821   44699850                       6   \n",
      "39748  chr17   66358140   66358465                       5   \n",
      "41178  chr17   81563970   81564002                       5   \n",
      "41703  chr17   83255944   83255985                       5   \n",
      "43256  chr19     417865     417884                       5   \n",
      "44055  chr19    1735672    1735729                       5   \n",
      "44273  chr19    2313065    2313104                       5   \n",
      "45392  chr19    6185555    6185572                       6   \n",
      "46625  chr19   13721617   13721668                       5   \n",
      "47694  chr19   19067936   19068035                       5   \n",
      "47786  chr19   19624891   19624944                       5   \n",
      "51011  chr19   52894666   52894707                       5   \n",
      "51070  chr19   53313329   53313384                       6   \n",
      "56845   chr2  218277148  218277174                       5   \n",
      "58423  chr20    1006128    1006173                       5   \n",
      "58494  chr20    1862302    1862338                       5   \n",
      "58676  chr20    4114523    4114637                       5   \n",
      "71794   chr4    1613126    1613150                       5   \n",
      "78610   chr6   33505369   33505425                       5   \n",
      "81170   chr7    1270900    1270925                       6   \n",
      "81238   chr7    1386485    1386510                       5   \n",
      "85986   chr8    1272322    1272363                       6   \n",
      "86099   chr8    2186250    2186281                       6   \n",
      "88987   chr8  144519924  144519973                       5   \n",
      "89244   chr8  145179779  145179807                       5   \n",
      "\n",
      "                                              haplotypes  \\\n",
      "93     [HG02587_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "114    [HG02953_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "140    [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "160    [HG02953_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "165    [HG02953_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "1335   [HG02554_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "1352   [HG02282_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "8879   [HG02554_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "13227  [HG03065_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "13842  [HG02587_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "18724  [HG02282_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "18746  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "26188  [HG02587_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "30169  [HG02282_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "35165  [HG02554_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "35210  [HG02666_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "35812  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "38078  [HG02666_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "38465  [HG02282_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "39748  [HG02554_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "41178  [HG02011_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "41703  [HG02666_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "43256  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "44055  [HG02282_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "44273  [HG02554_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "45392  [HG02011_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "46625  [HG03452_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "47694  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "47786  [HG02282_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "51011  [HG02587_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "51070  [HG02282_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "56845  [HG01890_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "58423  [HG02011_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "58494  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "58676  [HG03520_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "71794  [HG03371_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "78610  [HG02587_hap1_quadron_filtered_verkko_batch1_p...   \n",
      "81170  [HG02011_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "81238  [HG01890_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "85986  [HG02011_hap1_quadron_filtered_verkko_batch2_p...   \n",
      "86099  [HG02666_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "88987  [HG02666_hap2_quadron_filtered_verkko_batch1_p...   \n",
      "89244  [HG02011_hap2_quadron_filtered_verkko_batch2_p...   \n",
      "\n",
      "                                        superpopulations  \\\n",
      "93     [African Ancestry, African Ancestry, African A...   \n",
      "114    [African Ancestry, African Ancestry, African A...   \n",
      "140    [African Ancestry, African Ancestry, African A...   \n",
      "160    [African Ancestry, African Ancestry, African A...   \n",
      "165    [African Ancestry, African Ancestry, African A...   \n",
      "1335   [African Ancestry, African Ancestry, African A...   \n",
      "1352   [African Ancestry, African Ancestry, African A...   \n",
      "8879   [African Ancestry, African Ancestry, African A...   \n",
      "13227  [African Ancestry, African Ancestry, African A...   \n",
      "13842  [African Ancestry, African Ancestry, African A...   \n",
      "18724  [African Ancestry, African Ancestry, African A...   \n",
      "18746  [African Ancestry, African Ancestry, African A...   \n",
      "26188  [African Ancestry, African Ancestry, African A...   \n",
      "30169  [African Ancestry, African Ancestry, African A...   \n",
      "35165  [African Ancestry, African Ancestry, African A...   \n",
      "35210  [African Ancestry, African Ancestry, African A...   \n",
      "35812  [African Ancestry, African Ancestry, African A...   \n",
      "38078  [African Ancestry, African Ancestry, African A...   \n",
      "38465  [African Ancestry, African Ancestry, African A...   \n",
      "39748  [African Ancestry, African Ancestry, African A...   \n",
      "41178  [African Ancestry, African Ancestry, African A...   \n",
      "41703  [African Ancestry, African Ancestry, African A...   \n",
      "43256  [African Ancestry, African Ancestry, African A...   \n",
      "44055  [African Ancestry, African Ancestry, African A...   \n",
      "44273  [African Ancestry, African Ancestry, African A...   \n",
      "45392  [African Ancestry, African Ancestry, African A...   \n",
      "46625  [African Ancestry, African Ancestry, African A...   \n",
      "47694  [African Ancestry, African Ancestry, African A...   \n",
      "47786  [African Ancestry, African Ancestry, African A...   \n",
      "51011  [African Ancestry, African Ancestry, African A...   \n",
      "51070  [African Ancestry, African Ancestry, African A...   \n",
      "56845  [African Ancestry, African Ancestry, African A...   \n",
      "58423  [African Ancestry, African Ancestry, African A...   \n",
      "58494  [African Ancestry, African Ancestry, African A...   \n",
      "58676  [African Ancestry, African Ancestry, African A...   \n",
      "71794  [African Ancestry, African Ancestry, African A...   \n",
      "78610  [African Ancestry, African Ancestry, African A...   \n",
      "81170  [African Ancestry, African Ancestry, African A...   \n",
      "81238  [African Ancestry, African Ancestry, African A...   \n",
      "85986  [African Ancestry, African Ancestry, African A...   \n",
      "86099  [African Ancestry, African Ancestry, African A...   \n",
      "88987  [African Ancestry, African Ancestry, African A...   \n",
      "89244  [African Ancestry, African Ancestry, African A...   \n",
      "\n",
      "                               superpopulation_codes  unique_sample_count  \n",
      "93     [AFR, AFR, AFR, AFR, AFR, AFR, AFR, AFR, AFR]                    7  \n",
      "114                   [AFR, AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "140                        [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "160                        [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "165                        [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "1335                       [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "1352                       [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "8879                       [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "13227                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "13842                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "18724                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "18746                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "26188                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "30169                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "35165                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "35210                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "35812                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "38078                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "38465                 [AFR, AFR, AFR, AFR, AFR, AFR]                    6  \n",
      "39748                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "41178                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "41703                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "43256                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "44055                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "44273                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "45392                 [AFR, AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "46625                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "47694                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "47786                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "51011                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "51070                 [AFR, AFR, AFR, AFR, AFR, AFR]                    6  \n",
      "56845                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "58423                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "58494                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "58676                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "71794                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "78610                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "81170                 [AFR, AFR, AFR, AFR, AFR, AFR]                    6  \n",
      "81238                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "85986                 [AFR, AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "86099                 [AFR, AFR, AFR, AFR, AFR, AFR]                    6  \n",
      "88987                      [AFR, AFR, AFR, AFR, AFR]                    5  \n",
      "89244                      [AFR, AFR, AFR, AFR, AFR]                    5  \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "def extract_sample_name(filename):\n",
    "    \"\"\"Extract the first part of the filename before the first underscore.\"\"\"\n",
    "    return filename.split('_')[0]\n",
    "\n",
    "def process_intersections(bed_file, metadata_file):\n",
    "    \"\"\"Process BED file and merge it with metadata to find ancestry-specific regions.\"\"\"\n",
    "    # Read the BED file into a DataFrame\n",
    "    df = pd.read_csv(bed_file, sep='\\t', header=None, \n",
    "                     names=['chrom', 'start', 'end', 'score', 'filename', 'chrom2', 'start2', 'end2', 'score2'])\n",
    "    \n",
    "    # Drop duplicate filenames for the same region\n",
    "    df_unique = df.drop_duplicates(subset=['chrom', 'start', 'end', 'filename']).copy()\n",
    "    \n",
    "    # Extract sample name from filename\n",
    "    df_unique['Sample'] = df_unique['filename'].apply(extract_sample_name)\n",
    "    \n",
    "    # Read the metadata file into a DataFrame\n",
    "    metadata_df = pd.read_csv(metadata_file, delimiter='\\t')\n",
    "    \n",
    "    # Merge metadata with the DataFrame\n",
    "    df_merged = pd.merge(df_unique, \n",
    "                         metadata_df[['Sample name', 'Sex', 'Superpopulation name', 'Superpopulation code']],\n",
    "                         left_on='Sample', right_on='Sample name', how='left')\n",
    "    \n",
    "    # Group by the chrom, start, and end columns and aggregate unique filenames and metadata\n",
    "    grouped = df_merged.groupby(['chrom', 'start', 'end']).agg(\n",
    "        unique_haplotype_count=('filename', 'nunique'),\n",
    "        haplotypes=('filename', lambda x: list(x)),\n",
    "        superpopulations=('Superpopulation name', lambda x: list(x)),\n",
    "        superpopulation_codes=('Superpopulation code', lambda x: list(x)),\n",
    "        unique_sample_count=('Sample', 'nunique')\n",
    "    ).reset_index()\n",
    "    \n",
    "    return grouped\n",
    "\n",
    "def find_shared_regions(grouped_intersections):\n",
    "    \"\"\"Identify regions shared by 5 or more haplotypes of the same ancestry.\"\"\"\n",
    "    shared_regions = grouped_intersections[\n",
    "        (grouped_intersections['superpopulations'].apply(lambda x: len(set(x)) == 1)) & \n",
    "        (grouped_intersections['unique_sample_count'] >= 5)\n",
    "    ]\n",
    "    return shared_regions\n",
    "\n",
    "# Main\n",
    "bed_file_path = '/home/alextu/scratch/results/bed_files/chm13/quadron_haplotype_intersections_chm13_query/verkko_batch123_chm13_quadron_posstrand_intersected_output.bed'\n",
    "metadata_file_path = '/home/alextu/scratch/igsr_sample_metadata/igsr_samples.tsv'\n",
    "\n",
    "# Process intersections\n",
    "grouped_intersections = process_intersections(bed_file_path, metadata_file_path)\n",
    "\n",
    "# Find regions shared by 5 or more haplotypes of the same ancestry\n",
    "shared_regions = find_shared_regions(grouped_intersections)\n",
    "\n",
    "# Print the shared regions\n",
    "print(shared_regions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb2370e5-b51f-4be2-be4c-59cf6c30a9fa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.11",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
